Initial PR for new template: forest with two domains

wkasdorp opened the pull request:
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use resourceGroup().location for resource locations
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

- [ ] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

wkasdorp commented:
This template builds a forest with 2 domains, which many enterprises have. It is highly configurable and illustrates some new features such as Managed Disks, and the use of the "condition" template element.

wkasdorp commented:
apologies for the accidental closure of the PR ...

bmoore-msft commented on 301-create-ad-forest-with-subdomain/README.md:
```diff
@@ -0,0 +1,147 @@
+although I did not fully succeed in this. For repeatable jobs I use 
+subtemplates. Creating a new VM is a nice example. 
+
+In the octobe 2017 update I greatly simplified the use
```
October

wkasdorp replied:
fixed.

bmoore-msft commented on 301-create-ad-forest-with-subdomain/README.md:
```diff
@@ -0,0 +1,147 @@
+This is almost undocumented, but the short version is that almost
+ nothing worked for 2008 R2 so I had to give it up. 
+
+### Update october 2017
```
capitalize October

wkasdorp replied:
fixed

bmoore-msft commented on 301-create-ad-forest-with-subdomain/azuredeploy.json:
```diff
@@ -0,0 +1,711 @@
+      "metadata": {
+        "description": "The name of the admin account for the Domain(s)"
+      },
+      "defaultValue": "adAdministrator",
```
remove default value - increases surface area

wkasdorp replied:
done.

bmoore-msft commented on 301-create-ad-forest-with-subdomain/azuredeploy.json:
```diff
@@ -0,0 +1,711 @@
+
+    "VMSize": {
+      "type": "string",
+      "allowedValues": [
```
remove allowed values unless there is a specific req for the size

wkasdorp replied:
There _is_ a dependency on size. Anything smaller does not deploy reliably (too slow), anything larger has no point. Also, in a subordinate template I parse the VM type name to see what kind of managed disk (Standard/Premium) is needed.

Still want to me to take the allowed values out?

bmoore-msft commented on 301-create-ad-forest-with-subdomain/azuredeploy.parameters.json:
```diff
@@ -0,0 +1,21 @@
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "adminUsername": {
+      "value": "adAdministrator"
```
GEN-UNIQUE

wkasdorp replied:
thx, missed that one.

bmoore-msft commented on 301-create-ad-forest-with-subdomain/azuredeploy.parameters.json:
```diff
@@ -0,0 +1,21 @@
+    "adminPassword": {
+      "value": "GEN-PASSWORD"
+    },
+    "VMSize": {
```
remove the params that have default values

wkasdorp replied:
done.

bmoore-msft commented on 301-create-ad-forest-with-subdomain/nested/CreateAndPrepnewVM.json:
```diff
@@ -0,0 +1,219 @@
+      "metadata": {
+        "description": "The name of the admin account for the VM"
+      },
+      "defaultValue": "adAdministrator"
```
remove default value

wkasdorp replied:
done.

bmoore-msft commented on 301-create-ad-forest-with-subdomain/nested/CreateAndPrepnewVM.json:
```diff
@@ -0,0 +1,219 @@
+      "properties": {
+        "publicIPAllocationMethod": "Dynamic",
+        "dnsSettings": {
+          "domainNameLabel": "[concat(parameters('VMName'),'-',uniqueString(resourceGroup().id))]"
```
vmname can chars/casing not supported in a dns name - we need to account for those or just use uniqueString

wkasdorp replied:
wow, nice catch!

bmoore-msft commented on 301-create-ad-forest-with-subdomain/nested/CreateAndPrepnewVM.json:
```diff
@@ -0,0 +1,219 @@
+      }
+    },
+    {
+      "type": "Microsoft.Compute/disks",
```
can we just use implicit disks?

wkasdorp replied:
there is an implicit managed disk for the OS, but this is an explicit managed disk for the AD database with a specific size and properties. So no, I don't think we can make it implicit.

bmoore-msft replied:
Everything we're doing here can be done using and implicit disk by just setting the same properties (size, createOption, storageAccountType) - let's go that route it will simplify the template.

bmoore-msft commented on 301-create-ad-forest-with-subdomain/nested/CreateAndPrepnewVM.json:
```diff
@@ -0,0 +1,219 @@
+            "typeHandlerVersion": "1.8",
+            "autoUpgradeMinorVersion": true,
+            "protectedSettings": {
+              "commandToExecute": "powershell.exe -ExecutionPolicy Unrestricted -encodedcommand cwBlAHQALQBlAHgAZQBjAHUAdABpAG8AbgBwAG8AbABpAGMAeQAgAHUAbgByAGUAcwB0AHIAaQBjAHQAZQBkAA=="
```
does the user ever need to changes this?

wkasdorp replied:
No. it is a base64 encode string to set the Powershell execution mode to unrestricted.

wkasdorp commented:
That's excellent Brian, I'll get on it!

wkasdorp commented:
Hi Brian,

I'm done making the requested changes. I fixed one additional item: the "deploy to azure" button was pointing to my private repo.

bmoore-msft commented:
Let's move to an implicit disk for the AD data disk - everything else looks good.

wkasdorp commented:
Hi Brian, 

I added the implicit data disk. It automatically picks up the correct storage SKU, which was the one thing I was worried about. 

You may see that the previous Travis run failed; not sure why this happened, but perhaps it was a timeout. I changed no code between the current successful run and the previous failure. All I did was change README.md to trigger another Travis run...

wkasdorp commented:
Brian, do you think we are OK now?

wkasdorp commented:
Hi Brian, tested it again with its final configuration on GitHub, and it still works fine. Thanks for the work that you put into this.
